File: tools/content/review-variants.ts
import { readFile } from 'node:fs/promises';
import { z } from 'zod';
import { reviewQuestion, type AuthoredReviewTemplate } from './review-templates.js';
import { validateMath } from './content.js';

const text = z.string().trim().min(1);
const byte = z.number().int().min(0).max(31);
const schema = z.record(
  text,
  z
    .object({
      selection: z
        .object({
          hashBytes: z.array(byte).min(1),
          moduli: z.array(z.number().int().min(1).max(256)).min(1),
          choiceRotationByte: byte.optional(),
        })
        .strict(),
      variants: z
        .array(
          z
            .object({
              id: text,
              parameters: z.record(text, z.number().int()),
              question: reviewQuestion,
            })
            .strict(),
        )
        .min(1),
    })
    .strict(),
);
export type ReviewVariants = z.infer<typeof schema>;

// Every question and parameter value is authored data. This validates the finite
// table and its historical seed lookup, without constructing curriculum.
export function validateReviewVariants(raw: unknown, templates: AuthoredReviewTemplate[]) {
  const banks = schema.parse(raw);
  const expected = templates.filter((t) => t.family === 'generated');
  if (Object.keys(banks).length !== expected.length || expected.some((t) => !banks[t.id]))
    throw Error('Review variant coverage mismatch');
  for (const template of expected) {
    const bank = banks[template.id];
    const { hashBytes, moduli, choiceRotationByte } = bank.selection;
    if (hashBytes.length !== moduli.length || new Set(hashBytes).size !== hashBytes.length)
      throw Error('Invalid review variant selection: ' + template.id);
    if (moduli.reduce((n, m) => n * m, 1) !== bank.variants.length)
      throw Error('Incomplete review variant table: ' + template.id);
    for (const [index, variant] of bank.variants.entries()) {
      let remaining = index;
      const digits = moduli.map(() => 0);
      for (let i = moduli.length - 1; i >= 0; i--) {
        digits[i] = remaining % moduli[i];
        remaining = Math.floor(remaining / moduli[i]);
      }
      if (variant.id !== digits.join(':'))
        throw Error('Review variant identity/order mismatch: ' + template.id);
      const q = variant.question;
      if (/\{\{[^}]+\}\}/.test(JSON.stringify(q)))
        throw Error('Review variants must contain complete questions: ' + template.id);
      if (q.choice && q.assessment) throw Error('Conflicting review grading methods');
      if (q.choice) {
        const ids = q.choice.options.map((o) => o.id);
        if (new Set(ids).size !== ids.length || !ids.includes(q.choice.correctOption))
          throw Error('Invalid review variant choices: ' + template.id);
        if (template.evidenceLevel !== 'recognition')
          throw Error('Choices require recognition evidence: ' + template.id);
      } else if (choiceRotationByte !== undefined) {
        throw Error('Review choice rotation requires choices: ' + template.id);
      }
      for (const value of [
        q.instructions,
        q.prompt,
        q.answer,
        q.math || '',
        ...(q.choice?.options.flatMap((o) => [o.text, o.feedback]) || []),
      ])
        validateMath(value);
    }
  }
  // Preserve authored key ordering for source review digests.
  return raw as ReviewVariants;
}
export async function loadReviewVariants(templates: AuthoredReviewTemplate[]) {
  return validateReviewVariants(
    JSON.parse(await readFile('content/review-variants.json', 'utf8')),
    templates,
  );
}

File: server/review_variants.go
package main

import (
	"encoding/json"
	"fmt"
	"strconv"
	"strings"
)

// Reject incomplete or malformed tables at startup, before a persisted seed can
// select a different question or panic. Curriculum is supplied by the catalog.
func validateReviewVariantBanks(c Catalog) error {
	expected := map[string]bool{}
	for _, template := range c.ReviewTemplates {
		if template.Generator == "" {
			continue
		}
		expected[template.ID] = true
		bank, ok := c.ReviewVariants[template.ID]
		if !ok {
			return fmt.Errorf("Missing review variant bank: %s", template.ID)
		}
		selection := bank.Selection
		if len(selection.HashBytes) == 0 || len(selection.HashBytes) != len(selection.Moduli) {
			return fmt.Errorf("Invalid review variant selection: %s", template.ID)
		}
		count := 1
		used := map[int]bool{}
		for i, modulus := range selection.Moduli {
			b := selection.HashBytes[i]
			if b < 0 || b >= 32 || used[b] || modulus < 1 || modulus > 256 || count > len(bank.Variants)/modulus {
				return fmt.Errorf("Invalid review variant selection: %s", template.ID)
			}
			used[b] = true
			count *= modulus
		}
		if count != len(bank.Variants) {
			return fmt.Errorf("Incomplete review variant bank: %s", template.ID)
		}
		if b := selection.ChoiceRotationByte; b != nil && (*b < 0 || *b >= 32) {
			return fmt.Errorf("Invalid review choice rotation: %s", template.ID)
		}
		for index, variant := range bank.Variants {
			digits := make([]string, len(selection.Moduli))
			remaining := index
			for i := len(digits) - 1; i >= 0; i-- {
				digits[i] = strconv.Itoa(remaining % selection.Moduli[i])
				remaining /= selection.Moduli[i]
			}
			if variant.ID != strings.Join(digits, ":") || variant.Question == nil || variant.Parameters == nil {
				return fmt.Errorf("Invalid review variant identity/data: %s", template.ID)
			}
			raw, err := json.Marshal(variant.Question)
			if err != nil {
				return err
			}
			var q struct {
				Choice     *ChoiceAssessment
				Assessment *Assessment
			}
			if err = json.Unmarshal(raw, &q); err != nil {
				return err
			}
			if q.Choice != nil && q.Assessment != nil {
				return fmt.Errorf("Conflicting review variant grading: %s", template.ID)
			}
			if q.Assessment != nil {
				if err = validateAssessment(q.Assessment); err != nil {
					return err
				}
			}
			if q.Choice != nil {
				seen := map[string]bool{}
				for _, option := range q.Choice.Options {
					if option.ID == "" || seen[option.ID] {
						return fmt.Errorf("Invalid review variant choices: %s", template.ID)
					}
					seen[option.ID] = true
				}
				if len(seen) < 2 || !seen[q.Choice.CorrectOption] {
					return fmt.Errorf("Invalid review variant answer: %s", template.ID)
				}
			} else if selection.ChoiceRotationByte != nil {
				return fmt.Errorf("Review choice rotation requires choices: %s", template.ID)
			}
		}
	}
	for id := range c.ReviewVariants {
		if !expected[id] {
			return fmt.Errorf("Unknown review variant bank: %s", id)
		}
	}
	return nil
}
